devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -0,0 +1,277 @@
+# MetGenC Pipeline Refactoring Plan
+
+## Overview
+
+This document outlines a plan to incrementally refactor the MetGenC pipeline to separate decision-making from execution, implement immutable data structures, and disentangle reading/writing from processing components.
+
+## Current Pipeline Issues
+
+The current pipeline has several architectural characteristics:
+
+1. **Mixed concerns**: Decision-making and execution are intertwined throughout the operations
+2. **Embedded I/O**: Reading and writing occur within pipeline operations rather than at boundaries
+3. **Mutable state**: The `Granule` dataclass is modified throughout the pipeline
+4. **Tight coupling**: Components depend heavily on each other and configuration state
+
+## Key Benefit Goals
+
+1. **Separation of Concerns**: Decision-making is isolated from execution
+2. **Immutable Data**: All data structures are immutable, preventing side effects
+3. **Testability**: Each phase can be tested independently with predictable inputs/outputs
+4. **Maintainability**: Clear boundaries between reading, processing, building, and writing
+5. **Flexibility**: New readers, processors, or outputs can be added without affecting other phases
+6. **Error Handling**: Failures are contained within phases and can be handled appropriately
+
+## Success Criteria
+
+- [ ] All I/O operations occur only at pipeline boundaries
+- [ ] Decision-making is separated from execution
+- [ ] All data structures are immutable
+- [ ] Each pipeline phase can be tested independently
+- [ ] Existing functionality is preserved
+- [ ] Performance is maintained or improved
+- [ ] Code coverage remains above 90%
+
+## Risk Mitigation
+
+1. **Feature Flags**: Toggle between old/new implementations -- either through CLI or .INI
+2. **Gradual Migration**: One component / feature at a time
+3. **Comprehensive Testing**: Test each change thoroughly
+4. **Performance Monitoring**: Track pipeline performance throughout migration
+
+This refactoring plan ensures the pipeline becomes more maintainable, testable, and extensible while minimizing risk through incremental changes that preserve existing functionality.
+
+## New Pipeline Architecture
+
+### Pipeline Diagram
+
+```
+┌─────────────────────────────────────────────────────────────────┐
+│                    DECISION PHASE                               │
+├─────────────────────────────────────────────────────────────────┤
+│ 1. Configuration Analysis                                       │
+│    └─ Parse config → Determine processing strategy              │
+│                                                                 │
+│ 2. File Discovery & Grouping                                   │
+│    └─ Scan directories → Group related files                   │
+│                                                                 │
+│ 3. Reader Selection                                             │
+│    └─ Analyze file types → Select appropriate readers          │
```

Comment:
There may be different "readers" needed to process a single dataset, because there can be different sources of temporal, spatial, and miscellaneous UMM-G/CNM content. Source could be data file content, ini file content, data file name, or premet/spatial/spo files.